drivers/gnss/gnss_quectel_lx6.c

```diff
@@ -0,0 +1,732 @@
+{
+	const struct quectel_lx6_config *config = dev->config;
+	struct quectel_lx6_data *data = dev->data;
+	int ret;
+
+	ret = modem_chat_run_script(&data->chat, &full_cold_start_script);
+	if (ret < 0) {
+		LOG_ERR("Failed to full cold restart GNSS: %d", ret);
+		modem_pipe_close(data->pipe, K_SECONDS(10));
+		return ret;
+	}
+
+	if (config->i2c_bus) {
+		/* Closing pipe while LX6 is not responsive to I2C commands */
+		modem_pipe_close(data->pipe, K_SECONDS(10));
+
+		k_sleep(K_SECONDS(1));
+
+		ret = modem_pipe_open(data->pipe, K_SECONDS(10));
+		if (ret < 0) {
+			LOG_ERR("Failed to open modem pipe: %d", ret);
+			return ret;
+		}
+	} else {
+		k_sleep(K_SECONDS(1));
+	}
+
+	return 0;
+}
+
+static int quectel_lx6_resume(const struct device *dev)
+{
+	const struct quectel_lx6_config *config = dev->config;
+	struct quectel_lx6_data *data = dev->data;
+	int ret;
+
+	LOG_DBG("Resume");
+
+	if (config->vcc.port != NULL) {
+		ret = gpio_pin_set_dt(&config->vcc, 1);
+		if (ret < 0) {
+			LOG_ERR("Failed to set VCC high: %d", ret);
+			return ret;
+		}
+		k_msleep(250);
+	} else {
+		/* TODO add check that either VCC or FORCE_ON are defined */
+		return -ENOTSUP;
+	}
+
+	ret = modem_pipe_open(data->pipe, K_SECONDS(10));
+	if (ret < 0) {
+		LOG_ERR("Failed to open modem pipe: %d", ret);
+		return ret;
+	}
+
+	ret = modem_chat_attach(&data->chat, data->pipe);
+	if (ret < 0) {
+		LOG_ERR("Failed to attach chat: %d", ret);
+		modem_pipe_close(data->pipe, K_SECONDS(10));
+		return ret;
+	}
+
+	ret = quectel_lx6_configure_pps(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to configure PPS: %d", ret);
+		modem_pipe_close(data->pipe, K_SECONDS(10));
+		return ret;
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int quectel_lx6_suspend(const struct device *dev)
+{
+	const struct quectel_lx6_config *config = dev->config;
+	struct quectel_lx6_data *data = dev->data;
+	int ret;
+
+	LOG_DBG("Suspend");
+
+	if (config->vcc.port != NULL) {
+		ret = gpio_pin_set_dt(&config->vcc, 0);
+		if (ret < 0) {
+			LOG_ERR("[%s] VCC pin active", dev->name);
+		}
+	} else {
+		/* TODO add check that either VCC or FORCE_ON are defined */
+		return -ENOTSUP;
+	}
+
+	modem_pipe_close(data->pipe, K_SECONDS(10));
+
+	return ret;
+}
+
+static int quectel_lx6_turn_on(const struct device *dev)
+{
+	struct quectel_lx6_data *data = dev->data;
+	int ret;
+
+	LOG_INF("Turn on");
+
+	k_msleep(250);
+
+	ret = modem_pipe_open(data->pipe, K_SECONDS(10));
+	if (ret < 0) {
+		LOG_ERR("Failed to open modem pipe: %d", ret);
+	}
+
+	return ret;
+}
+
+static int quectel_lx6_turn_off(const struct device *dev)
+{
+	struct quectel_lx6_data *data = dev->data;
+
+	LOG_INF("Turn off");
+
+	return modem_pipe_close(data->pipe, K_SECONDS(10));
+}
+
+static int quectel_lx6_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	struct quectel_lx6_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	(void)k_mutex_lock(&data->mutex, K_FOREVER);
```

Comment:
Please use a k_sem for locking instead of a mutex :) The mutex is quite a bit more costly than a semaphore :)
```
/* init */
k_sem_init(&sem, 1, 1);
/* lock */
(void)k_sem_take(&sem, K_FOREVER);
/*unlock*/
k_sem_give(&sem);